hydra-node/src/Hydra/Logging/Monitoring.hs

```diff
@@ -5,30 +5,35 @@ module Hydra.Logging.Monitoring (
 
 import Hydra.Prelude
 
+import Control.Monad.Class.MonadSTM (MonadSTM (readTVarIO), modifyTVar', newTVarIO)
 import Control.Tracer (Tracer (Tracer))
 import Data.Map.Strict as Map
+import Hydra.HeadLogic (Effect (ClientEffect), Event (NetworkEvent), HydraMessage (ReqTx), ServerOutput (SnapshotConfirmed), Snapshot (confirmed))
+import Hydra.Ledger (Tx (TxId), txId)
 import Hydra.Logging.Messages (HydraLog (..))
 import Hydra.Network (PortNumber)
-import Hydra.Node (HydraNodeLog (ProcessedEvent))
+import Hydra.Node (HydraNodeLog (ProcessedEffect, ProcessedEvent, ProcessingEvent))
 import System.Metrics.Prometheus.Http.Scrape (serveMetrics)
-import System.Metrics.Prometheus.Metric (Metric (CounterMetric))
-import System.Metrics.Prometheus.Metric.Counter (inc)
+import System.Metrics.Prometheus.Metric (Metric (CounterMetric, HistogramMetric))
+import System.Metrics.Prometheus.Metric.Counter (add, inc)
+import System.Metrics.Prometheus.Metric.Histogram (observe)
 import System.Metrics.Prometheus.MetricId (Name (Name))
-import System.Metrics.Prometheus.Registry (Registry, new, registerCounter, sample)
+import System.Metrics.Prometheus.Registry (Registry, new, registerCounter, registerHistogram, sample)
 
 -- | Wraps a monadic action using a `Tracer` and capture metrics based on traces.
 -- Given a `portNumber`, this wrapper starts a Prometheus-compliant server on this port.
 -- This is a no-op if given `Nothing`. This function is not polymorphic over the type of
 -- messages because it needs to understand them in order to provide meaningful metrics.
 withMonitoring ::
-  (MonadIO m, MonadAsync m) =>
+  (MonadIO m, MonadAsync m, Tx tx, MonadMonotonicTime m) =>
   Maybe PortNumber ->
   Tracer m (HydraLog tx net) ->
   (Tracer m (HydraLog tx net) -> m ()) ->
   m ()
 withMonitoring Nothing tracer action = action tracer
 withMonitoring (Just monitoringPort) (Tracer tracer) action = do
-  (traceMetric, registry) <- prepareRegistry
+  txMap <- newTVarIO mempty
+  (traceMetric, registry) <- prepareRegistry txMap
```

Comment:
`txMap` is not used outside of `prepareRegistry`? If yes, you should create it in there.